Extend `.Bind()` to Support TypedBinding

brminnick opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Markup Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Markup Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI Markup core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui.markup/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

 ### Description of Change ###

This PR extends the `.Bind()` extension method by providing the option of using [`TypedBinding`](https://github.com/dotnet/maui/blob/64e517538f179919cd27ae7e4a380d47b2541904/src/Controls/src/Core/TypedBinding.cs#L71) which is the binding engine used by XAML Compiled Bindings to improve performance and ensure Type safety. 

```cs
// One-way (aka read-only) Binding
new Label().Row(Row.Description).Bind(Label.TextProperty, static (StoryModel m) => m.Description)

// Two-way Binding
new Entry().Bind(Entry.TextProperty, static (SettingsViewModel vm) => vm.NumberOfTopStoriesToFetch, static (SettingsViewModel vm, int text) => vm.NumberOfTopStoriesToFetch = text)
```

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes #155 

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui.Markup/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/204


 ### Additional information ###

 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->

brminnick commented:
@danielftz Here's the link to the NuGet Package generated by this PR that contains this `TypedBinding` support: https://dev.azure.com/dotnet/696bc9fd-f160-4e97-a1bd-7cbbb3b58f66/_apis/build/builds/83190/artifacts?artifactName=nuget&api-version=7.0&%24format=zip

danielftz commented:
> > Note: Proposal #155 is not yet approved. I opened this Draft PR to produce a NuGet Package that the community can use to demo the new API. Added the #149 label until #155 is Approved.
> 
> ### Description of Change
> This PR extends the `.Bind()` extension method by providing the option of using [`TypedBinding`](https://github.com/dotnet/maui/blob/64e517538f179919cd27ae7e4a380d47b2541904/src/Controls/src/Core/TypedBinding.cs#L71) which is the binding engine used by XAML Compiled Bindings to improve performance and ensure Type safety.
> 
> ```cs
> // One-way (aka read-only) Binding
> new Label().Row(Row.Description).Bind(Label.TextProperty, (StoryModel m) => m.Description)
> 
> // Two-way Binding
> new Entry().Bind(Entry.TextProperty, (SettingsViewModel vm) => vm.NumberOfTopStoriesToFetch, (SettingsViewModel vm, int text) => vm.NumberOfTopStoriesToFetch = text)
> ```
> 
> ### Linked Issues
> * Fixes [[Proposal] Typed Bindings #155](https://github.com/CommunityToolkit/Maui.Markup/issues/155)
> 
> ### PR Checklist
> * [x]  Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
> * [ ]  Has tests (if omitted, state reason in description)
> * [x]  Has samples (if omitted, state reason in description)
> * [x]  Rebased on top of `main` at time of PR
> * [x]  Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui.Markup/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
> * [ ]  Documentation created or updated
> 
> ### Additional information
From what I understand of compiled binding, it would throw build error when the binding is invalid. 
e.g. 
```
// One-way (aka read-only) Binding, error expected
new Label().Row(Row.Description).Bind(Label.TextProperty, (StoryModel m) => m.ThisNotAString)
```
Could this be done somehow?

danielftz commented:
Also OneWay typed binding does not seem to work, demo project [here](https://github.com/danielftz/TypedBindingBug)

Platform: Android API33

brminnick commented:
Good catch, @danielftz! I'll update the PR with it now that the team approved the Proposal!

brminnick commented:
### BindingMode.OneWay

> Also OneWay typed binding does not seem to work

Yea, I wonder if this is a bug in .NET MAUI?  

I removed the logic that was previously assigning `BindingMode.OneWay` when `setter is null` and updated the `mode` parameter's default value, `BindingMode mode = BindingMode.Default`.

### Handlers

> Turns out the Tuple<Func<TBindingContext, object>, string>[] handlers parameter of the TypedBinding object isn't useless after all

I figured out that we can retrieve the property name from the `getter` by changing its type to an `Expression`, `Expression<Func<TBindingContext, TSource>> getter`.

With an expression, we can retrieve the Func's property name like so:
```cs
static string GetMemberName<T>(in Expression<T> expression) => expression.Body switch
{
	MemberExpression m => m.Member.Name,
	UnaryExpression u when u.Operand is MemberExpression m => m.Member.Name,
	_ => throw new InvalidOperationException("Could not retreive member name")
};
```

This allows us to keep the original method syntax without requiring the user to also input the property name:
```cs
new Label().Row(Row.Description).Bind(Label.TextProperty, (StoryModel m) => m.Description)
```

### Updated NuGet Package

Here's the latest NuGet Package generated from this PR if you'd like to try it out: https://dev.azure.com/dotnet/696bc9fd-f160-4e97-a1bd-7cbbb3b58f66/_apis/build/builds/83653/artifacts?artifactName=nuget&api-version=7.0&%24format=zip

danielftz commented:
Thank you very much! 

From what Ive seen in the [TypedBinding Unit Test](https://github.com/dotnet/maui/blob/64e517538f179919cd27ae7e4a380d47b2541904/src/Controls/tests/Core.UnitTests/TypedBindingUnitTests.cs#L87) , the `handler` parameter is an array. I think there should be a way to parse the expression in a way that should allow for binding to a more complex path, which turns out to be a pretty common use case no my project. 

e.g. 
```
new Label().Row(Row.Description).Bind(Label.TextProperty, (StoryViewModel vm) => vm.Service.Model.Description);
```

brminnick commented:
>  I think there should be a way to parse the expression in a way that should allow for binding to a more complex path

@danielftz Good call! 

I've added an overloaded method to `.Bind()` that adds a parameter for 		`(Func<TBindingContext, object?>, string)[] handlers`.

### Base `.Bind()` Extension Methods for Typed Bindings

Here's what the two base methods for typed binding looks like now:

```cs
        // For Simple Bindings

	public static TBindable Bind<TBindable, TBindingContext, TSource, TParam, TDest>(
		this TBindable bindable,
		BindableProperty targetProperty,
		Expression<Func<TBindingContext, TSource>> getter,
		Action<TBindingContext, TSource>? setter = null,
		BindingMode mode = BindingMode.Default,
		Func<TSource?, TParam?, TDest>? convert = null,
		Func<TDest?, TParam?, TSource>? convertBack = null,
		TParam? converterParameter = default,
		string? stringFormat = null,
		TBindingContext? source = default,
		TDest? targetNullValue = default,
		TDest? fallbackValue = default) where TBindable : BindableObject);

        // For Complex Bindings
	public static TBindable Bind<TBindable, TBindingContext, TSource, TParam, TDest>(
		this TBindable bindable,
		BindableProperty targetProperty,
		Func<TBindingContext, TSource> getter, // Uses a `Func` because handlers are provided and don't need to be inferred
		(Func<TBindingContext, object?>, string)[] handlers,
		Action<TBindingContext, TSource>? setter = null,
		BindingMode mode = BindingMode.Default,
		Func<TSource?, TParam?, TDest>? convert = null,
		Func<TDest?, TParam?, TSource>? convertBack = null,
		TParam? converterParameter = default,
		string? stringFormat = null,
		TBindingContext? source = default,
		TDest? targetNullValue = default,
		TDest? fallbackValue = default) where TBindable : BindableObject);
```

### Usage Example

This allows us to keep the simple syntax for simple bindings while still providing an overloaded method for more-complex bindings:

```cs
// Simple Binding
var slider = new Slider().Bind<Slider, ViewModel, double>(Slider.ValueProperty, static (ViewModel viewModel) => viewModel.Percentage, static (ViewModel viewModel, double temperature) => viewModel.Percentage = temperature);

// Complex Binding
var entry = new Entry().Bind(Entry.TextColorProperty,
							static (NestedViewModel vm) => vm.Model?.Model?.TextColor,
							new (Func<NestedViewModel, object?>, string)[]
							{
								(vm => vm, "Model"),
								(vm => vm.Model, "Model"),
								(vm => vm.Model.Model, "Text")
							},
							static (NestedViewModel vm, Color? color) =>
							{
								if (vm.Model?.Model?.TextColor is not null && color is not null)
								{
									vm.Model.Model.TextColor = color;
								}
							},
							bindingMode);

brminnick commented:
@danielftz Just following up - have you had a chance to try out the [latest iteration of this PR's NuGet Package](https://github.com/CommunityToolkit/Maui.Markup/pull/156#issuecomment-1340089862) in your app to verify the `handlers` implementation works? 

I'm really liking how it turned out, architecturally, and I've verified it working in a few of my apps. But before we begin writing the documentation, I'd love for you to verify it as well!

danielftz commented:
> @danielftz Just following up - have you had a chance to try out the [latest iteration of this PR's NuGet Package](https://github.com/CommunityToolkit/Maui.Markup/pull/156#issuecomment-1340089862) in your app to verify the `handlers` implementation works?
> 
> I'm really liking how it turned out, architecturally, and I've verified it working in a few of my apps. But before we begin writing the documentation, I'd love for you to verify it as well!

Thank you for the hard work! Seems all good to me

brminnick commented:
Great! Thanks @danielftz.

@bijington - FYI, I'll get started on the TypedBinding documentation this week!

bijington reviewed (approved):
Looks good to me!

brminnick commented:
/azp run

danielftz commented:
Is there any way to do two way binding while still supporting conversion? 

e.g. two way binding between `Entry.TextProperty` and a `double` value.

brminnick commented on src/CommunityToolkit.Maui.Markup.UnitTests/TypedBindingExtensionsTests.cs:
```diff
@@ -0,0 +1,410 @@
+		if (setContextFirst)
+		{
+			entry.BindingContext = viewmodel;
+			entry.Bind(Entry.TextColorProperty,
+							static (NestedViewModel vm) => vm.Model?.Model?.TextColor,
+							new (Func<NestedViewModel, object?>, string)[]
+							{
+								(vm => vm, nameof(NestedViewModel.Model)),
+								(vm => vm.Model, nameof(NestedViewModel.Model)),
+								(vm => vm.Model.Model, nameof(NestedViewModel.Model.TextColor))
+							},
+							static (NestedViewModel vm, Color? color) =>
+							{
+								if (vm.Model?.Model?.TextColor is not null && color is not null)
+								{
+									vm.Model.Model.TextColor = color;
+								}
+							},
+							bindingMode);
```
@bijington Here's an example of how to use TypedBindings for nested bindings.

E.g. If there is a Property in the ViewModel that has a Property inside it (like using a Model as a Property in the ViewModel)

brminnick commented:
> Is there any way to do two way binding while still supporting conversion?
@danielftz Yup! You just add a converter using the `convert` and `convertBack` parameters:

```cs
new Entry().Bind(Entry.TextProperty, static (ViewModel) vm => vm.Number, static (ViewModel vm, double number) => vm.Number = number, convert: (double number) => number.ToString(), convertBack: (string text) => double.Parse(text));
```

bijington reviewed (approved):
Just my point on the Bind with no setter and BindingMode option but that probably still works

brminnick commented on src/CommunityToolkit.Maui.Markup/TypedBindingExtensions.cs:
```diff
@@ -0,0 +1,200 @@
+		BindableProperty targetProperty,
+		Expression<Func<TBindingContext, TSource>> getter,
+		Action<TBindingContext, TSource>? setter = null,
+		BindingMode mode = BindingMode.Default,
```
Hmmm...yea, I guess that would be a valid input.

I'll add a unit test today to test it out. Depending on how it does/doesn't work, I may add an `ArgumentException`.

```cs
if(setter is not null && mode is not BindingMode.TwoWay)
{
  throw new ArgumentException("...");
}
```

`BindingMode.Default` is `BindingMode.TwoWay` for the controls that support it, so I should also check the control's default binding too

bijington replied:
Or maybe we just accept that the consumer knows what they are doing around this area? Bindings let you set them up incorrectly as they currently stand.

brminnick replied:
Hey Shaun! I was just playing around with this argument validation idea, and after giving it some thought, I think it's best for us to pass through the user's inputs to .NET MAUI.

If the .NET MAUI team was to do argument validation, they're welcome to, but in this case, since `BindableObject.SetBinding()` allows these combination of values to be set, we should honor that.

Plus, if we do argument validation, then we risk a breaking change to `CommunityToolkit.Maui.Markup` whenever the .NET MAUI team updates the binding engine.

For posterity, here's the code I wrote to validate the inputs if we ever change our mind and decide to add it to our TypedBindings:

```cs
if (setter is not null)
{
	if (mode is not BindingMode.TwoWay or BindingMode.OneWayToSource)
	{
		throw new ArgumentException($"""
Invalid {nameof(BindingMode)}.
A {nameof(BindingMode)} of {mode} cannot be used to set the value of a property on the binding source (typically a View Model).
Either remove the {nameof(setter)} (changing it to null), or set {nameof(BindingMode)} {nameof(mode)} to either {nameof(BindingMode.TwoWay)} or {nameof(BindingMode.OneWayToSource)}
""", nameof(mode));
	}

	if (mode is BindingMode.Default && targetProperty.DefaultBindingMode is not BindingMode.TwoWay or BindingMode.OneWayToSource)
	{
		throw new ArgumentException($"""
A {nameof(setter)} cannot be applied because the {nameof(BindingMode.Default)} is {targetProperty.DefaultBindingMode}.
Either remove the {nameof(setter)} (changing it to null), or change {nameof(BindingMode)} {nameof(mode)} to either {nameof(BindingMode.TwoWay)} or {nameof(BindingMode.OneWayToSource)}
""", nameof(mode));
	}
}
```
